### PR TITLE
Symbolic dimensions allowed in MultivariateEwens

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References to other Issues or PRs

[1] Extends https://github.com/sympy/sympy/pull/16576

#### Brief description of what is fixed or changed
Symbolic dimensions can now be passed to `MultivariateEwens` in `sympy.stats.joint_rv_types`.

#### Other comments
This is a work in progress task. There are some problems with `ArrayComprehension` which need to be handled, for instance, it is not possible to multiply the elements of `ArrayComprehension` and return an unevaluated `Mul` object. 

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by czgdp1807** [comment]
ping @Upabjojr @sidhantnagpal @kangzhiq

**Comment by kangzhiq** [comment]
```
>>> M = Mul.fromiter(ArrayComprehension(i, (i, 1, n)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\CodeHouse\Python\sympy\sympy\core\basic.py", line 259, in fromiter
    return cls(*tuple(args), **assumptions)
TypeError: 'ArrayComprehension' object is not iterable
```

So more precisely, do you mean that `ArrayComprehension` should be iterable?

**Comment by czgdp1807** [comment]
@kangzhiq As you would have seen the diff, I am able to get the things done without, `ArrayComprehension` but the logic is separate for integer and symbolic dimensions.
Basically, `ArrayComprehension` should be an iterable and whenever operators like `*` or `fromiter` are used it should remain unevaluated if dimensions are symbolic and `Mul`, `Add` should also remain unevaluated.

**Comment by oscarbenjamin** [comment]
I don't see how ArrayComprehension can remain unevaluated with `*` and fromiter. They both require an actual iterable and there's no way a symbolic ArrayComprehension can do that. There could be a way to construct a symbolic Product from a symbolic ArrayComprehension though.

**Comment by czgdp1807** [comment]
> There could be a way to construct a symbolic Product from a symbolic ArrayComprehension though.

I don't think we should create an `ArrayComprehension` to just convert it into `Product`, instead we can directly create the `Product` object and this is what I have done in the `diff`.

**Comment by kangzhiq** [comment]
As far as I can understand,  what you are expecting from `ArrayComprehension` that multiplying each term by iterating on a list of function is supposed to be handled by `Product`.  
However, when I try to test it by using:
```
j, k = symbols('j, k', positive=True, integer=True)
term_2 = Product(theta**syms[j]/((j+1)**syms[j]*factorial(syms[j])),(j, 0, n - 1)).doit()
cond = Eq(Sum((k + 1)*syms[k], (k, 0, n - 1)).doit(), n)
return Piecewise((term_1 * term_2, cond), (0, True))
```
so that you won't need to check the type of `n`, it continuously raises some errors. Have you already tried it?

**Comment by czgdp1807** [comment]
@kangzhiq I will try to unify the logic today and will update you accordingly if the things work correctly. Currently, I don't expect them to because there will be a problem of indexing the `syms`.

File: sympy/stats/tests/test_joint_rv.py [hunk 1]
```diff
@@ -150,7 +153,14 @@ def test_MultivariateEwens():
...
     raises(ValueError, lambda: MultivariateEwens('e1', 5, theta_f))
-    raises(ValueError, lambda: MultivariateEwens('e1', n, theta))
+
+    # tests for symbolic dimensions
+    eds = MultivariateEwens('E', n, theta)
+    a = IndexedBase('a')
+    den = ("Piecewise((factorial(n)*Product(theta**a[j]*(j + 1)**(-a[j])/"
+           "factorial(a[j]), (j, 0, n - 1))/RisingFactorial(theta, n),"
+           " Eq(n, Sum((k + 1)*a[k], (k, 0, n - 1)))), (0, True))")
+    assert str(density(eds)(a)) == den
```
Comment (Upabjojr) [inline sympy/stats/tests/test_joint_rv.py]: have you given a try to `dummy_eq`?

Comment (czgdp1807) [reply sympy/stats/tests/test_joint_rv.py]: Thanks it worked fine.

**Comment by czgdp1807** [comment]
> Currently, I don't expect them to because there will be a problem of indexing the `syms`.

This will cause a problem for symbolic `n`. Hence, the logic for symbolic dimensions and integer dimensions cannot be unified. 
The reason is for symbolic `n`, in `syms[j]`, `j` will be a symbol. So for combining the logic, we will be required to use `IndexedBase` because it allows both symbolic and integer indexes. However, I don't support it because it will not allow the use of API like, `density(ed)(x, y, z)`. The user will be restricted to use, `density(ed)(i)`, i is `IndexedBase` object which will lead to bad user experience and may cause a lot of problems. If @Upabjojr agrees to such restriction of arguments then I will make the change.

**Comment by Upabjojr** [comment]
I think the code looks good. Tell me if you're OK to merge.

**Comment by czgdp1807** [comment]
@Upabjojr Yes, please merge it.

**Pull request merged**